### PR TITLE
[Feat] 2/n Add Fault Tolerant Training to LightningFetcher

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds fault tolerant support to the LightningFetcher.

Blocked by https://github.com/PyTorchLightning/pytorch-lightning/pull/8890

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -39,11 +39,16 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Fault-tolerant training:
     * Added `FastForwardSampler` and `CaptureIterableDataset` injection to data loading utilities ([#8366](https://github.com/PyTorchLightning/pytorch-lightning/pull/8366))
+<<<<<<< HEAD
```
Comment (justusschock) [inline CHANGELOG.md]: bad merge

**Review by justusschock — APPROVED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (tchaton) [reply CHANGELOG.md]: Removed :) thanks

File: pytorch_lightning/utilities/apply_func.py [hunk 2]
```diff
@@ -77,7 +77,6 @@ def apply_to_collection(
 ) -> Any:
     """
     Recursively applies a function to all elements of a certain dtype.
-
```
Comment (awaelchli) [inline pytorch_lightning/utilities/apply_func.py:80]: @tchaton it's really essential that we keep these empty lines for proper rendering of docs

File: pytorch_lightning/utilities/auto_restart.py [hunk 3]
```diff
@@ -49,6 +51,7 @@ def __getattr__(self, key: str) -> Any:
         return getattr(self._sampler, key, None)
 
     def setup(self, dataloader_batch_size: Optional[int] = None) -> None:
+        # TODO: ask @tchaton about this docstring
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:54]: what's not good with this docstring?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py [hunk 4]
```diff
@@ -188,22 +299,29 @@ def _wrap_generator_samplers(self) -> None:
...
-    def reset_on_epoch(self) -> None:
+        self.reset_on_epoch()
+
+    def reset_on_epoch(self):
         self._state_dict = None
 
     def __iter__(self) -> Iterator:
         # create a generator from the wrapped Iterative Dataset
-        # if the dataset contained samplers, they will be transformers into generators
+        # if the dataset contained samplers, they will be transformeed into generators
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:315]:
```suggestion
        # if the dataset contained samplers, they will be transformed into generators
```

File: pytorch_lightning/utilities/apply_func.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/apply_func.py]: Yes, sounds good.

File: pytorch_lightning/utilities/apply_func.py [hunk 5]
```diff
@@ -95,6 +95,11 @@ def apply_to_collection(
     if isinstance(data, dtype) and (wrong_dtype is None or not isinstance(data, wrong_dtype)):
         return function(data, *args, **kwargs)
 
+    # range is a type implemented in C and is afaik the only sequence-like
+    # that does not accept other sequences for construction
+    if isinstance(data, range):
+        data = tuple(data)
```
Comment (Borda) [inline pytorch_lightning/utilities/apply_func.py:101]: maybe `list(data)`?

File: pytorch_lightning/utilities/auto_restart.py [hunk 6]
```diff
@@ -116,13 +125,114 @@ def _compute_current_iteration(self, num_batches_processed: Optional[int] = None
...
+    num_batches_fetched: int = 0
+    name: Optional[str] = None
+
+    @classmethod
+    def load_state_dict(cls, state_dict) -> "IteratorState":
+        return cls(**state_dict)
+
+
+@dataclass
+class CollectionIteratorState:
```
Comment (Borda) [inline pytorch_lightning/utilities/auto_restart.py:147]: why collection, the docstring does not talk about multiple...

File: pytorch_lightning/utilities/auto_restart.py [hunk 7]
```diff
@@ -116,13 +125,114 @@ def _compute_current_iteration(self, num_batches_processed: Optional[int] = None
...
+    represent_map_dataset: Optional[bool] = None
+
+    def update(self, iter_name: Optional[str], new_state: IteratorState) -> None:
+        self.represent_map_dataset = iter_name is None
+        if self.represent_map_dataset:
+            state = self.state
+        else:
+            if iter_name not in self.state:
+                self.state[iter_name] = {}
+            state = self.state[iter_name]
```
Comment (Borda) [inline pytorch_lightning/utilities/auto_restart.py:161]: how about if `iter_name` is None?

File: pytorch_lightning/utilities/auto_restart.py [hunk 8]
```diff
@@ -116,13 +125,114 @@ def _compute_current_iteration(self, num_batches_processed: Optional[int] = None
...
+
+        lastest_worker_id = new_state.worker_id
+        state[lastest_worker_id] = new_state
+        self.lastest_worker_id = lastest_worker_id
+
+    @property
+    def sampler_states(self) -> Dict[int, Any]:
+        return {0: self.state[k].sampler_state[0] for k in self.state.keys()}
```
Comment (Borda) [inline pytorch_lightning/utilities/auto_restart.py:169]: if all is with key 0, it means only the last is kept?
```suggestion
        return {k: self.state[k].sampler_state[0] for k in self.state.keys()}
```

File: pytorch_lightning/utilities/auto_restart.py [hunk 9]
```diff
@@ -116,13 +125,114 @@ def _compute_current_iteration(self, num_batches_processed: Optional[int] = None
...
+
+    @property
+    def dataset_states(self) -> Dict[int, Any]:
+        return {k: self.state[k].dataset_state[k] for k in self.state.keys()}
```
Comment (Borda) [inline pytorch_lightning/utilities/auto_restart.py:173]: lest add some docstring

File: pytorch_lightning/utilities/auto_restart.py [hunk 10]
```diff
@@ -116,13 +125,114 @@ def _compute_current_iteration(self, num_batches_processed: Optional[int] = None
...
+    @property
+    def worker_id(self) -> int:
+        worker_info = get_worker_info()
+        return worker_info.id if worker_info else 0
+
+    def __getitem__(self, item) -> Tuple[Any, Dict[int, Dict]]:
+        if self._cached_state_dict is not None:
+            if self.worker_id in self._cached_state_dict:
+                # TODO: reset random states
+                pass
```
Comment (Borda) [inline pytorch_lightning/utilities/auto_restart.py:210]: does this do anything?

Comment (awaelchli) [reply pytorch_lightning/utilities/auto_restart.py:210]: not yet, but in a follow up PR :) #8950

File: pytorch_lightning/utilities/auto_restart.py (hunk 7 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/auto_restart.py:161]: if itername is none, then the above if clause applies, since `if self.represent_map_dataset = iter_name is None = True`

File: pytorch_lightning/utilities/apply_func.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/apply_func.py]: sorry, this code shouldn't be here. it slipped in from the larger PR. removed!

File: pytorch_lightning/utilities/auto_restart.py [hunk 11]
```diff
@@ -72,6 +76,10 @@ def __iter__(self) -> Iterator[Any]:
             i += 1
 
         # here: i == self._current_iteration
+        if self._cached_state_dict is not None:
+            self._cached_state_dict = None
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/auto_restart.py:80]:
```suggestion
```
You are setting `self._cached_state_dict = None` in line 91 as well.

File: pytorch_lightning/utilities/auto_restart.py [hunk 12]
```diff
@@ -80,12 +88,13 @@ def __iter__(self) -> Iterator[Any]:
...
 
         self._current_iteration = 0
+        self._cached_state_dict = None
         self.restarting = False
 
     def __len__(self) -> int:
         return len(self._sampler)
 
-    def state_dict(self, num_batches_processed: Optional[int] = None) -> Dict[int, Dict[str, int]]:
+    def state_dict(self, num_batches_processed: Optional[int] = None) -> Dict[int, Dict[str, Any]]:
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/auto_restart.py:97]: Why change the type hint from int?

File: pytorch_lightning/utilities/auto_restart.py [hunk 13]
```diff
@@ -116,13 +125,114 @@ def _compute_current_iteration(self, num_batches_processed: Optional[int] = None
...
+    def load_state_dict(cls, state_dict) -> "IteratorState":
+        return cls(**state_dict)
+
+
+@dataclass
+class CollectionIteratorState:
+    """This class is used to hold the current iterator state and lives on the iterator."""
+
+    state: Union[Dict[Union[int, str], Union[Dict[str, IteratorState], IteratorState]]] = field(default_factory=dict)
+    lastest_worker_id: int = 0
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/auto_restart.py:151]:
```suggestion
    latest_worker_id: int = 0
```

File: pytorch_lightning/utilities/auto_restart.py [hunk 14]
```diff
@@ -188,22 +299,29 @@ def _wrap_generator_samplers(self) -> None:
...
     def __iter__(self) -> Iterator:
         # create a generator from the wrapped Iterative Dataset
-        # if the dataset contained samplers, they will be transformers into generators
+        # if the dataset contained samplers, they will be transformed into generators
         self.iter_data = iter(self.dataset)
 
         # wrap any generator associated to a Sampler into a `FastForwardSampler`.
+        if isinstance(self.iter_data, Generator):
+            raise MisconfigurationException(
+                "PyTorch Lightning Fault Tolerant does not support `__iter__` returning a generator."
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/auto_restart.py:321]:
```suggestion
                "PyTorch Lightning Fault-Tolerant feature does not support `__iter__` returning a generator."
```

File: pytorch_lightning/utilities/auto_restart.py (hunk 11 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/auto_restart.py:80]: true, but will be needed later when adding the random state stuff.
I will re-introduce it in the follow up PR. thanks

File: pytorch_lightning/utilities/auto_restart.py (hunk 12 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/auto_restart.py:95]: artifact from a larger PR where this will contain more than int in the future.

Comment (awaelchli) [reply pytorch_lightning/utilities/auto_restart.py:95]:
```suggestion
    def state_dict(self, num_batches_processed: Optional[int] = None) -> Dict[int, Dict[str, int]]:
```

**Review by ethanwharris — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py (hunk 8 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/auto_restart.py]: these methods are currently unused, I removed them. 
I will double check and address your comments in the follow up PR! Thanks

File: pytorch_lightning/utilities/auto_restart.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/auto_restart.py:145]: @tchaton could you help answer this question? :)

Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:145]: Yes, I called it collection but this isn t exactly accurate. It is more a merged iterator state.

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/auto_restart.py]: renamed and added docstring to make its usage clearer.
